### PR TITLE
Miscellanous issues found in Spidermonkey

### PR DESCRIPTION
**Pull request by eqrion**
Now this should be just about everything :)

### CODE REVIEW

File: crates/wast/src/ast/expr.rs [hunk 1]
```diff
@@ -399,9 +399,9 @@ instructions! {
         ElemDrop(ast::Index<'a>) : [0xfc, 0x0d] : "elem.drop",
         TableInit(TableInit<'a>) : [0xfc, 0x0c] : "table.init",
         TableCopy(TableCopy<'a>) : [0xfc, 0x0e] : "table.copy",
-        TableFill(ast::Index<'a>) : [0xfc, 0x11] : "table.fill",
-        TableSize(ast::Index<'a>) : [0xfc, 0x10] : "table.size",
-        TableGrow(ast::Index<'a>) : [0xfc, 0x0f] : "table.grow",
+        TableFill(TableArg<'a>) : [0xfc, 0x11] : "table.fill",
+        TableSize(TableArg<'a>) : [0xfc, 0x10] : "table.size",
+        TableGrow(TableArg<'a>) : [0xfc, 0x0f] : "table.grow",
```
Comment (alexcrichton) [inline crates/wast/src/ast/expr.rs]: Could some tests be added for this syntax where the immediate is optional?

File: crates/wast/src/binary.rs [hunk 2]
```diff
@@ -335,15 +335,13 @@ impl Encode for Limits {
...
-            self.limits.max.unwrap_or(0).encode(e);
-        } else {
-            self.limits.encode(e);
+        let flag_max = self.limits.max.is_some() as u8;
+        let flag_shared = self.shared as u8;
+        let flags = flag_max | (flag_shared << 1);
+        e.push(flags);
+        self.limits.min.encode(e);
+        if let Some(max) = self.limits.max {
+            max.encode(e);
```
Comment (alexcrichton) [inline crates/wast/src/binary.rs]: Out of curiosity, what was the motivator for this change?

File: crates/wast/src/resolve/names.rs [hunk 3]
```diff
@@ -111,6 +111,9 @@ impl<'a> Resolver<'a> {
                     ImportKind::Func(t) | ImportKind::Event(EventType::Exception(t)) => {
                         self.resolve_type_use(i.span, t)?;
                     }
+                    ImportKind::Global(t) => {
+                        self.resolve_valtype(&mut t.ty)?;
```
Comment (alexcrichton) [inline crates/wast/src/resolve/names.rs]: Could a test be added which would fail if this weren't here?

File: crates/wast/src/binary.rs (hunk 2 above)
Comment (eqrion) [reply crates/wast/src/binary.rs]: There's a test that constructs a `(memory 1 shared)` and asserts that the error message is that the maximum length is required for a shared memory. Before this change what would emit a maximum memory of 0, which would still fail but as a 'maximum must be greater than minimum error'. I thought the previous error message was more intuitive, so I added this change. It's a small thing, not very important.

[1] https://searchfox.org/mozilla-central/rev/278046367dab878316f60f0bd7f740cf73f3c447/js/src/jit-test/tests/wasm/memory-sharing.js#43

File: crates/wast/src/ast/expr.rs (hunk 1 above)
Comment (eqrion) [reply crates/wast/src/ast/expr.rs]: Yup, done!

File: crates/wast/src/resolve/names.rs (hunk 3 above)
Comment (eqrion) [reply crates/wast/src/resolve/names.rs]: Yup, done!

**Comment by alexcrichton** [comment]
👍

**Comment by eqrion** [comment]
Thanks!

**Comment by alexcrichton** [comment]
Oh also, want me to publish this as well?

**Comment by eqrion** [comment]
No, I'm probably going to need to write another patch for #69 first.

**Pull request merged**